src/tokens/ERC20Vault.sol

```diff
@@ -0,0 +1,161 @@
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice The underlying token the Vault accepts.
+    ERC20 public immutable underlying;
+
+    /// @notice The base unit of the underlying token and hence vault.
+    /// @dev Equal to 10 ** decimals. Used for fixed point arithmetic.
+    uint256 public immutable baseUnit;
+
+    /// @notice Creates a new Vault that accepts a specific underlying token.
+    /// @param _underlying The ERC20 compliant token the Vault should accept.
+    /// @param _namePrefix A name prefix to be applied before
+    /// @param _symbolPrefix The ERC20 compliant token the Vault should accept.
+
+    constructor(
+        ERC20 _underlying,
+        string memory _namePrefix,
+        string memory _symbolPrefix
+    )
+        ERC20(
+            string(abi.encodePacked(_namePrefix, _underlying.name(), " Vault")),
+            string(abi.encodePacked(_symbolPrefix, _underlying.symbol())),
+            _underlying.decimals()
+        )
+    {
+        underlying = _underlying;
+
+        baseUnit = 10**decimals;
+
+        // Prevent minting of shares until
+        // the initialize function is called.
+        totalSupply = type(uint256).max;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        DEPOSIT/WITHDRAWAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Emitted after a successful deposit.
+    /// @param user The address that deposited into the Vault.
+    /// @param underlyingAmount The amount of underlying tokens that were deposited.
+    event Deposit(address indexed user, uint256 underlyingAmount);
+
+    /// @notice Emitted after a successful withdrawal.
+    /// @param owner The address that withdrew from the Vault.
+    /// @param to The destination for withdrawn tokens.
+    /// @param underlyingAmount The amount of underlying tokens that were withdrawn.
+    event Withdraw(address indexed owner, address indexed to, uint256 underlyingAmount);
+
+    /// @notice Deposit a specific amount of underlying tokens.
+    /// @param to The address to receive shares corresponding to the deposit
+    /// @param underlyingAmount The amount of the underlying token to deposit.
+    function deposit(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        // We don't allow depositing 0 to prevent emitting a useless event.
+        require(underlyingAmount != 0, "AMOUNT_CANNOT_BE_ZERO");
+
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+        // Determine the equivalent amount of shares and mint them.
+        _mint(to, shares);
+
+        emit Deposit(to, underlyingAmount);
+
+        // Transfer in underlying tokens from the user.
+        // This will revert if the user does not have the amount specified.
+        underlying.safeTransferFrom(msg.sender, address(this), underlyingAmount);
+
+        afterDeposit(underlyingAmount);
```

Comment:
worries me doing this after an external call

Reply:
the intention behind this is that it will probably also do another external call in deposit like staking in the MasterChef or something. I'm down to enforce that to be a separate function but I also like leaving optional hook here